src/check/check_types/unify.zig

```diff
@@ -1,153 +1,3590 @@
+        return self == .ok;
     }
+};
 
-    try unifyType(allocator, type_store, &result, first, first_type, second, second_type);
+/// A temporary unification context used to unify two type variables within a `Store`.
+///
+/// `Unifier` is created per unification call and:
+/// * Resolves and compresses type variables
+/// * Applies unification logic across all supported `Content` variants
+/// * Allocates new variables into the store (via `fresh`)
+/// * Writes unified structure back into the store (via `merge`)
+///
+/// It works in tandem with `Scratch`, which owns reusable memory buffers
+/// for intermediate field lists, shared record partitions, and fresh variables.
+///
+/// `Unifier` supports:
+/// * flexible and rigid type variables
+/// * type aliases
+/// * extensible records with row polymorphism
+/// * basic support for function, tuple, and number types
+///
+/// Callers are not expected to construct `Unifier`. Instead  call `unify(...)`.
+const Unifier = struct {
+    const Self = @This();
 
-    return result;
-}
+    types_store: *store.Store,
+    scratch: *Scratch,
 
-fn unifyType(
-    allocator: Allocator,
-    type_store: *Type.Store,
-    result: *UnificationResult,
-    first: Type.Idx,
-    first_type: *Type,
-    second: Type.Idx,
-    second_type: *Type,
-) !void {
-    _ = allocator;
+    /// Init a unifier
+    /// Caller owns the memory of the provided values
+    fn init(types_store: *store.Store, scratch: *Scratch) Self {
+        return .{ .types_store = types_store, .scratch = scratch };
+    }
 
-    switch (first_type.*) {
-        .bool => {
-            @panic("todo");
-        },
-        .apply => {
-            @panic("todo");
-        },
-        .str => {
-            @panic("todo");
-        },
-        .int => |i| {
-            switch (i) {
-                .u8 => @panic("todo"),
-                .i8 => @panic("todo"),
-                .u16 => @panic("todo"),
-                .i16 => @panic("todo"),
-                .u32 => @panic("todo"),
-                .i32 => @panic("todo"),
-                .u64 => @panic("todo"),
-                .i64 => @panic("todo"),
-                .u128 => @panic("todo"),
-                .i128 => @panic("todo"),
+    // merge
+
+    /// Link the variables & updated the content in the type_store
+    /// In the old compiler, this function was called "merge"
+    fn merge(self: *Self, vars: *const ResolvedVarDescs, new_content: Content) void {
+        self.types_store.union_(vars.a.var_, vars.b.var_, .{
+            .content = new_content,
+            .rank = Rank.min(vars.a.desc.rank, vars.b.desc.rank),
+            .mark = Mark.none,
+        });
+    }
+
+    /// Create a new type variable *in this pool*
+    fn fresh(self: *Self, vars: *const ResolvedVarDescs, new_content: Content) Var {
+        const var_ = self.types_store.register(.{
+            .content = new_content,
+            .rank = Rank.min(vars.a.desc.rank, vars.b.desc.rank),
```

Comment:
I'm pretty sure this should be the maximum rank in the entire current pool - but `vars` only includes the variables presently unified, right?